pkg/object/azure.go

```diff
@@ -21,71 +21,73 @@ package object
 
 import (
 	"fmt"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
 	"io"
 	"log"
+	"net"
 	"net/url"
 	"os"
 	"strings"
-	"time"
-
-	"github.com/Azure/azure-sdk-for-go/storage"
 )
 
 type wasb struct {
 	DefaultObjectStorage
-	container *storage.Container
+	container *azblob.ContainerClient
+	cName     string
 	marker    string
 }
 
 func (b *wasb) String() string {
-	return fmt.Sprintf("wasb://%s/", b.container.Name)
+	return fmt.Sprintf("wasb://%s/", b.cName)
 }
 
 func (b *wasb) Create() error {
-	_, err := b.container.CreateIfNotExists(&storage.CreateContainerOptions{})
+	_, err := b.container.Create(ctx, nil)
+	if strings.Contains(err.Error(), string(azblob.StorageErrorCodeContainerAlreadyExists)) {
+		return nil
+	}
 	return err
 }
 
 func (b *wasb) Head(key string) (Object, error) {
-	blob := b.container.GetBlobReference(key)
-	err := blob.GetProperties(nil)
+	properties, err := b.container.NewBlobClient(key).GetProperties(ctx, &azblob.GetBlobPropertiesOptions{})
 	if err != nil {
 		return nil, err
 	}
 
 	return &obj{
-		blob.Name,
-		blob.Properties.ContentLength,
-		time.Time(blob.Properties.LastModified),
-		strings.HasSuffix(blob.Name, "/"),
+		key,
+		*properties.ContentLength,
+		*properties.LastModified,
+		strings.HasSuffix(key, "/"),
 	}, nil
 }
 
 func (b *wasb) Get(key string, off, limit int64) (io.ReadCloser, error) {
-	blob := b.container.GetBlobReference(key)
-	var end int64
-	if limit > 0 {
-		end = off + limit - 1
-	}
-	return blob.GetRange(&storage.GetBlobRangeOptions{
-		Range: &storage.BlobRange{
-			Start: uint64(off),
-			End:   uint64(end),
-		},
-	})
+	download, err := b.container.NewBlockBlobClient(key).Download(ctx, &azblob.DownloadBlobOptions{Offset: &off, Count: &limit})
+	if err != nil {
+		return nil, err
+	}
+	return download.BlobDownloadResponse.RawResponse.Body, err
 }
 
 func (b *wasb) Put(key string, data io.Reader) error {
-	return b.container.GetBlobReference(key).CreateBlockBlobFromReader(data, nil)
+	_, err := b.container.NewBlockBlobClient(key).UploadStreamToBlockBlob(ctx, data, azblob.UploadStreamToBlockBlobOptions{})
+	return err
 }
 
 func (b *wasb) Copy(dst, src string) error {
-	uri := b.container.GetBlobReference(src).GetURL()
-	return b.container.GetBlobReference(dst).Copy(uri, nil)
+	_, err := b.container.NewBlockBlobClient(dst).CopyFromURL(ctx, b.container.NewBlockBlobClient(src).URL(),
+		&azblob.CopyBlockBlobFromURLOptions{})
+	return err
 }
 
 func (b *wasb) Delete(key string) error {
-	return b.container.GetBlobReference(key).Delete(nil)
+	_, err := b.container.NewBlockBlobClient(key).Delete(ctx, &azblob.DeleteBlobOptions{})
+	if err != nil && strings.Contains(err.Error(), string(azblob.StorageErrorCodeBlobNotFound)) {
```

Comment:
Can we use errors.Is(err, azblob.StorageErrorCodeBlobNotFound)

Reply:
No, `azblob.StorageErrorCodeBlobNotFound` is string type, does not implement error interface.